src/Components/Web/src/DynamicRootComponentInterop.cs

```diff
@@ -0,0 +1,186 @@
+                var parameterName = parametersReader.GetString()!;
+                object? parameterValue;
+                if (TryGetComponentParameterType(componentType, parameterName, out var parameterType))
+                {
+                    // It's a statically-declared parameter, so we can parse it into a known .NET type
+                    parameterValue = JsonSerializer.Deserialize(
+                        ref parametersReader,
+                        parameterType,
+                        _jsonOptions);
+                }
+                else
+                {
+                    // Unknown parameter - possibly valid as "catch-all". Use whatever type appears
+                    // to be present in the JSON data.
+                    parametersReader.Read();
+                    switch (parametersReader.TokenType)
+                    {
+                        case JsonTokenType.Number:
+                            parameterValue = parametersReader.GetDouble();
+                            break;
+                        case JsonTokenType.String:
+                            parameterValue = parametersReader.GetString();
+                            break;
+                        case JsonTokenType.True:
+                        case JsonTokenType.False:
+                            parameterValue = parametersReader.GetBoolean();
+                            break;
+                        case JsonTokenType.Null:
+                            parameterValue = null;
+                            break;
+                        default:
+                            throw new ArgumentException($"There is no declared parameter named '{parameterName}', so the supplied object cannot be deserialized.");
+                    }
+                }
+
+                parameterViewBuilder.Add(parameterName, parameterValue);
+                parametersReader.Read();
+            }
+
+            return _renderer.RenderRootComponentAsync(componentId, parameterViewBuilder.ToParameterView());
+        }
+
+        /// <summary>
+        /// For framework use only.
+        /// </summary>
+        [JSInvokable]
+        public void RemoveRootComponent(int componentId)
+            => _renderer.RemoveRootComponent(componentId);
+
+        /// <inheritdoc />
+        public void Dispose()
+            => _selfReference.Dispose();
+
+        private bool TryGetComponentParameterType(Type componentType, string parameterName, out Type parameterType)
+        {
+            // The parameter name comes from untrusted input, and we don't want to hash arbitrarily long
+            // strings. This restriction is both to limit the impact of hostile input and to guide developers
+            // to pick parameter names that will keep the costs low.
+            if (parameterName.Length > 50)
+            {
+                throw new ArgumentException($"The parameter '{parameterName}' cannot be used dynamically because its name is too long.");
```

Comment:
Note: I'm going to remove this length limit. Discussed with @javiercn and it became clear that this serves no real purpose.